python/lbann/core/training_algorithm.py

```diff
@@ -0,0 +1,322 @@
+        return params
+
+class MetaLearningStrategy:
+    """Base class for metalearning strategies for LTFB."""
+    def __init__(self):
+        pass
+
+class LTFB(TrainingAlgorithm):
+    """Livermore tournament fast-batch (LTFB) algorithm.
+
+    This training algorithm is a simple composite training
+    algorithm. The MPI universe is subdivided into several
+    "trainers". Each trainer applies a local training algorithm. At
+    the completion of local training, a metaheuristic ("metalearning"
+    algorithm) is applied to select a new set of models to continue.
+
+    The usage requirements for this training algorithm are a
+    fully-specified local training algorithm and stopping criteria for
+    the outer loop.
+    """
+
+    class StoppingCriteria:
+        """Stopping criteria for LTFB"""
+        def __init__(self, metalearning_steps: int = 1):
+            """Construct a new LTFB stopping criteria object
+
+            Args:
+                metalearning_steps:
+                  The number of outer-loop iterations.
+            """
+            self.metalearning_steps = metalearning_steps
+
+        def export_proto(self):
+            """Get a protobuf representation of this object."""
+            msg = AlgoProto.LTFB.TerminationCriteria()
+            msg.max_tournaments = self.metalearning_steps
+            return msg
+
+    def __init__(self, name: str, local_algo: TrainingAlgorithm,
+                 metalearning: MetaLearningStrategy,
+                 metalearning_steps: int = 1):
+        """Construct a new LTFB algorithm.
+
+        Args:
+            name:
+              A user-defined name to identify this object in logs.
+            local_algo:
+              The trainer-local algorithm to apply.
+            metalearning:
+              The metalearning strategy to apply after local training.
+            metalearning_steps:
+              The number of outer-loop iterations.
+        """
+        self.name = name
+        self.metalearning = metalearning
+        self.local_algo = local_algo
+        self.stopping = self.StoppingCriteria(metalearning_steps=metalearning_steps)
+
+    def do_export_proto(self):
+        """Get a protobuf representation of this object."""
+        params = AlgoProto.LTFB()
+        params.stopping_criteria.CopyFrom(self.stopping.export_proto())
+        params.meta_learning_strategy.Pack(self.metalearning.export_proto())
+        params.local_training_algorithm.CopyFrom(self.local_algo.export_proto())
+        return params
+
+class RandomPairwiseExchange(MetaLearningStrategy):
+    """The classic LTFB pairwise tourament metalearning strategy.
+
+    This metalearning strategy is the original algorithm used for
+    LTFB. After each local training step, all trainers in the LBANN
+    environment pair off and have a "tournament" (if the number of
+    trainers is odd, one competes with itself). The tournament is a
+    simple comparison of a metric value that's read from the
```

Comment:
Arguably that will represent a new strategy. That's part of the point of this refactor/new direction -- the strategy pattern should isolate these decisions into smaller units that are (hopefully) easy to write and register with the factory. When we have a new idea, we just write a new strategy instead of adding a bunch of spaghetti code and confusing switches to shim it into the existing code.